packages/@react-stately/data/src/useAsyncList.ts

```diff
@@ -256,21 +307,27 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
   // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
+  let filteredItems = useMemo(
+    () =>  filter ? data.items.filter(item => filter(item, data.filterText)) : data.items,
+    [data.items, data.filterText, filter]);
+
   return {
-    items: data.items,
+    items: filteredItems,
     selectedKeys: data.selectedKeys,
     sortDescriptor: data.sortDescriptor,
-    isLoading: data.state === 'loading' || data.state === 'loadingMore' || data.state === 'sorting',
+    // TODO: add isFiltering? This is so we can have the loading icon in the textfield when filter text changes instead of in the listbox
+    isLoading: data.state === 'loading' || data.state === 'loadingMore' || data.state === 'sorting' || data.state === 'filtering',
     error: data.error,
+    filterText: data.filterText,
     getItem(key: Key) {
       return data.items.find(item => getKey(item) === key);
     },
     reload() {
       dispatchFetch({type: 'loading'}, load);
     },
     loadMore() {
-      // Ignore if already loading more.
-      if (data.state === 'loadingMore' || data.cursor == null) {
+      // Ignore if already loading more or if performing server side filtering.
+      if (data.state === 'loadingMore' || (data.state === 'filtering' && !filter) || data.cursor == null) {
```

Comment:
hmm what prompted this?

Reply:
This wasn't prompted by any kind of broken behavior, more just me thinking through the cases/states that could happen. I figured if server side filtering was happening (and was exceptionally slow), the user could trigger a `loadMore`  by scrolling through the existing items and hitting the bottom of the listbox. If that happens, I figured we should still wait for the server side filtering to finish rather than dispatching a new load more request (feels extraneous especially since we wouldn't have the proper up to date cursor)